Gutenboarding: update domain picker styling

andrewserong opened the pull request:
#### Changes proposed in this Pull Request

Quoting https://github.com/Automattic/wp-calypso/pull/40742#issuecomment-609659426:

> ![image](https://user-images.githubusercontent.com/14988353/78539535-12f78580-7836-11ea-87f8-82fb2cb169e9.png)
> 
> Update from today. I've done the following:
> 
> * [x] Restyled the suggestion items as radio input fields, with checked, focus, and hover states
> * [x] Updated the domain form to use `onDomainSelect` when selecting a paid domain instead of `onDomainPurchase` (this will require the rest of the domain flow to be in place as it means cutting off the current domain flow), but it ensures that selecting a domain works as expected
> * [x] Add a popover container so that the popover can be fixed in position matching the positioning in the Figma — this ensures that the popover doesn't move around as the user selects their domain and the text is updated in the header
> * [x] Added search icon
> * [x] Added close button
>
> There's still a bit to be tidied up — the spacing is approximate in places, so might need to be double-checked / tweaked against the Figma. And the spacing in the loading state needs to be adjusted, as it's current too short. I also haven't addressed any mobile styling.

(Note by @ockham -- I think we can tweak the above in a follow-up.)

#### Testing instructions

- http://calypso.localhost:3000/gutenboarding
- Verify that the domain picker looks as in the above screenshot, and that it still works.

andrewserong commented:
Note: this is a WIP for updating the domain picker styling to match the current Figma, there's a fair bit that hasn't been done yet, most notably adding in the radio buttons.

So far, the layout has enough room for everything if the user enters a short domain name:

![image](https://user-images.githubusercontent.com/14988353/78336803-b6009300-75db-11ea-8222-e39a82442032.png)

However with a long domain name the `Recommended` pill bumps up against the `Free` text and the whole area wraps.

![image](https://user-images.githubusercontent.com/14988353/78336976-f6601100-75db-11ea-8f3b-520e636ac973.png)

A few questions for @Automattic/gutenboarding-design —

* How should we handle long domain names like in this screenshot, when it's a recommended domain, to prevent the weird wrapping?
* The Figma uses a custom radio button. Does this already exist anywhere? If not, I'm happy to add in a couple of SVGs for the two states of the radio button / or do it in CSS instead. Is there any interaction / hover state to be aware of?
* What does the `More options` button do? And if it's to open up a full sized interface like in the Figma, can we leave it off until we've built the full interface for domains?
* There's an option for `I already own a domain` — what happens when a user selects this option? (The existing signup flow has a Transfer Domain sub-flow, which we won't be able to do in time for CfT, so how should we handle this option?)

simison commented:
I would try to stick to core radio button for consistency https://github.com/WordPress/gutenberg/tree/master/packages/components/src/radio-control

andrewserong commented:
> I would try to stick to core radio button for consistency https://github.com/WordPress/gutenberg/tree/master/packages/components/src/radio-control

Nice one, I'll give that a try on Monday!

simison commented:
Some more issues listed here; https://github.com/Automattic/wp-calypso/issues/40395

Let's close that one and create a new for any follow-ups once this is merged.

simison commented:
> Let's use radio buttons like the mock

Can you re-consider this? I wouldn't like us to introduce new primitive components, and prefer to stick to either Calypso components or WordPress components as much as possible. That'll make everything much easier down the road and more consistent overall. If the design mock is different from Calypso/Gutenberg component, we should change the design, or improve the component in core. :-) It's just a radio button, after all.

dubielzyk commented:
> It's just a radio button, after all.
Feel free to use the Calpyso radio component, just as long as visually it's a radio button. The reason for this is that radio buttons is a very known pattern. I don't think it's worth investing in making a bespoke solution when a perfectly great pattern exist already :)

andrewserong commented:
![image](https://user-images.githubusercontent.com/14988353/78539535-12f78580-7836-11ea-87f8-82fb2cb169e9.png)

Update from today. I've done the following:

* [x] Restyled the suggestion items as radio input fields, with checked, focus, and hover states
* [x] Updated the domain form to use `onDomainSelect` when selecting a paid domain instead of `onDomainPurchase` (this will require the rest of the domain flow to be in place as it means cutting off the current domain flow), but it ensures that selecting a domain works as expected
* [x] Add a popover container so that the popover can be fixed in position matching the positioning in the Figma — this ensures that the popover doesn't move around as the user selects their domain and the text is updated in the header
* [x] Added search icon
* [x] Added close button

@Automattic/luna — please feel free to jump in on this PR and change anything you like / or leave notes for me to address again tomorrow. There's still a bit to be tidied up — the spacing is approximate in places, so might need to be double-checked / tweaked against the Figma. And the spacing in the loading state needs to be adjusted, as it's current too short.

Note that the removal of the `onDomainPurchase` option is a breaking change, but hopefully this is consistent with the direction we're going in with keeping the user in Gutenboarding instead of popping out to buy a domain name at this stage.

Thanks!

ockham commented:
Thanks a lot @andrewserong! I'll take this over :smile:

ockham commented on client/landing/gutenboarding/components/domain-picker-button/index.tsx:
```diff
@@ -89,14 +88,24 @@ const DomainPickerButton: FunctionComponent< Props > = ( {
-						onDomainPurchase={ handlePaidDomainSelect }
+				<div className="domain-picker-button__popover-container">
+					<Popover
+						className="domain-picker-button"
```
This seems to belong to the button rather than the popover?

ockham replied:
It does collide with https://github.com/Automattic/wp-calypso/pull/40742/files#diff-9fb3f50c91d02f132bfab858e87cb792R70 :grimacing:

ockham commented on client/landing/gutenboarding/components/domain-picker/index.tsx:
```diff
@@ -59,12 +59,47 @@ export interface Props {
-	onClose,
-	currentDomain,
-} ) => {
+const FreeDomainIcon = () => (
+	<Icon
+		icon={ () => (
+			<svg
+				width="15"
+				height="15"
+				viewBox="0 0 15 15"
+				fill="none"
+				xmlns="http://www.w3.org/2000/svg"
+			>
+				<path
+					d="M13.2878 8.60833L8.61291 13.2844C8.4918 13.4057 8.34798 13.5019 8.18968 13.5676C8.03137 13.6332 7.86169 13.667 7.69032 13.667C7.51895 13.667 7.34926 13.6332 7.19096 13.5676C7.03265 13.5019 6.88884 13.4057 6.76773 13.2844L1.16699 7.68876V1.16699H7.68706L13.2878 6.76919C13.5307 7.01358 13.667 7.34417 13.667 7.68876C13.667 8.03335 13.5307 8.36395 13.2878 8.60833V8.60833Z"
+					stroke="#008A20"
+					strokeWidth="1.5"
+					strokeLinecap="round"
+					strokeLinejoin="round"
+				/>
+				<circle cx="4.50033" cy="4.50033" r="0.833333" fill="#008A20" />
+			</svg>
+		) }
+	/>
+);
```
Could maybe use the [tag icon](https://github.com/WordPress/gutenberg/blob/a97f31c4a80f4d945d10cfd6a7a5301baad9dabf/packages/icons/src/library/tag.js) from the `@wordpress/icons` npm

ockham replied:
Oh, Calypso isn't using the latest (G2 compliant) version yet :confused: -- it would look like this with icons from `@wordpress/icons`:

![image](https://user-images.githubusercontent.com/96308/78550557-913b3400-7804-11ea-9d99-7ac3537dd191.png)

This is even with `@wordpress/icons` v1.2.0 installed in `client/` -- unfortunately, v1.1.0 is still installed at root `node_modules/` level, and it seems like Gutenboarding is picking up the latter :confused: 

For reference, this is what it should look like:

![image](https://user-images.githubusercontent.com/96308/78554910-4a513c80-780c-11ea-9b24-89dbbbf764f4.png)

So we'll probably have to wait for #39644 to land before we can use `@wordpress/icons` v1.2.0 :confused:

ockham replied:
Filed #40803 to track this.

simison commented on client/landing/gutenboarding/components/domain-picker-button/style.scss:
```diff
@@ -7,3 +7,17 @@
+.domain-picker-button__popover-container {
+	position: fixed;
+	left: 72px;
+	top: 54px;
```
Doesn't the popover component itself take care of this? I sense trouble on mobile. :-)

ockham replied:
> Doesn't the popover component itself take care of this? I sense trouble on mobile. :-)

Andrew talks about this change [here](https://github.com/Automattic/wp-calypso/pull/40742#issuecomment-609659426):

> * [x] Add a popover container so that the popover can be fixed in position matching the positioning in the Figma — this ensures that the popover doesn't move around as the user selects their domain and the text is updated in the header

I agree with the mobile trouble tho :confused:

simison replied:
Aight, sounds like worth figuring out once we get to mobile designs later on. :-)

ockham commented on client/landing/gutenboarding/components/domain-picker/index.tsx:
```diff
@@ -59,12 +59,47 @@ export interface Props {
+	/>
+);
+
+const SearchIcon = () => (
+	<Icon
+		icon={ () => (
+			<svg
+				width="24"
+				height="24"
+				viewBox="0 0 24 24"
+				fill="none"
+				xmlns="http://www.w3.org/2000/svg"
+			>
+				<path d="M6 18L10 14.5" stroke="black" strokeWidth="1.5" />
+				<circle cx="13.5" cy="11.5" r="4.75" stroke="black" strokeWidth="1.5" />
+			</svg>
+		) }
+	/>
+);
```
Could maybe use the [search icon](https://github.com/WordPress/gutenberg/blob/a97f31c4a80f4d945d10cfd6a7a5301baad9dabf/packages/icons/src/library/search.js) from the `@wordpress/icons` npm

ockham replied:
https://github.com/Automattic/wp-calypso/pull/40742/files#r404025438
Tackled in #40803

ockham commented on client/landing/gutenboarding/components/close-button/index.tsx:
```diff
@@ -0,0 +1,29 @@
+	const { __: NO__ } = useI18n();
+	return (
+		<Button onClick={ onClose } label={ NO__( 'Close dialog' ) }>
+			<svg
+				width="12"
+				height="12"
+				viewBox="0 0 16 16"
+				fill="none"
+				xmlns="http://www.w3.org/2000/svg"
+			>
+				<path d="M1.40456 1L15 15M1 15L14.5954 1" stroke="#1E1E1E" strokeWidth="1.5" />
+			</svg>
```
Could maybe use the [tag icon](https://github.com/WordPress/gutenberg/blob/a97f31c4a80f4d945d10cfd6a7a5301baad9dabf/packages/icons/src/library/tag.js) from the `@wordpress/icons` npm

ockham replied:
https://github.com/Automattic/wp-calypso/pull/40742/files#r404025438

ockham commented:
I've polished this a bit. Setting to 'needs review' -- we can address some smaller bits in a follow-up.

ockham commented:
Rebased to fix e2e tests.

ramonjd commented on client/landing/gutenboarding/components/domain-picker/index.tsx:
```diff
@@ -84,12 +119,11 @@ const DomainPicker: FunctionComponent< Props > = ( {
-							{ NO__( 'Skip for now' ) }
-						</Button>
+						<div className="domain-picker__header-title">{ NO__( 'Choose a domain' ) }</div>
+						<CloseButton onClose={ () => onClose() } />
```
More a question than a review comment: is there any reason nowadays to worry about anonymous functions in props being an anti-pattern, or to prefer the following?

`<CloseButton onClose={ onClose } />`

Just asking for my own knowledge :)

andrewserong replied:
I think the latter should be fine since we don't need to pass anything into `onClose()` (that's the only reason I can think of in functional components for using an arrow function here).

andrewserong commented:
@ockham thanks for picking this one up, and for the feedback @simison!

I've pushed an update for dealing with the `position: fixed` styling in this PR to switch on the `expandOnMobile` prop for the `Popover` component on mobile, and only use this fixed positioning on viewports greater than `small`. Hopefully this should alleviate the main mobile issues until we can style it up properly for mobile. Here's how it looks now:

![domain-picker-mobile-small](https://user-images.githubusercontent.com/14988353/78623183-687d7200-78ca-11ea-856d-c287c39feac6.gif)

andrewserong commented:
Just noting that it appears the designs for the domain picker have changed slightly in the Figma (not sure if these are ready to be developed or still in progress?), so we can take a look at further styling separately. For now, I'll merge this one in.

razvanpapadopol commented:
@andrewserong This would be the follow-up implementing the updated designs: https://github.com/Automattic/wp-calypso/pull/40918

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/3198905

Thank you @andrewserong for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3198905).
